### PR TITLE
8281236: (D)TLS key exchange named groups

### PR DESCRIPTION
**Pull request by XueleiFan**
This update is to support key exchange named groups customization for individual (D)TLS connection. Please review the CSR as well:
CSR: https://bugs.openjdk.org/browse/JDK-8291950
RFE: https://bugs.openjdk.org/browse/JDK-8281236
Release-note: https://bugs.openjdk.org/browse/JDK-8291975

This is an effort similar to [JDK-8280494: "(D)TLS signature schemes"](https://bugs.openjdk.org/browse/JDK-8280494)

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8281236](https://bugs.openjdk.org/browse/JDK-8281236): (D)TLS key exchange named groups
 * [JDK-8291950](https://bugs.openjdk.org/browse/JDK-8291950): (D)TLS key exchange named groups (**CSR**)

### Reviewers
 * [Sean Mullan](https://openjdk.org/census#mullan) (@seanjmullan - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9776/head:pull/9776` \
`$ git checkout pull/9776`

Update a local copy of the PR: \
`$ git checkout pull/9776` \
`$ git pull https://git.openjdk.org/jdk pull/9776/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9776`

View PR using the GUI difftool: \
`$ git pr show -t 9776`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9776.diff">https://git.openjdk.org/jdk/pull/9776.diff</a>

</details>

### CODE REVIEW

**Comment by XueleiFan** [comment]
8281236: (D)TLS key exchange named groups

**Comment by XueleiFan** [comment]
Ping ...

**Comment by XueleiFan** [comment]
Anyone has cycle for the review?

**Comment by mcpowers** [comment]
Well crafted. LGTM.

**Comment by XueleiFan** [comment]
Thank you very much, @mcpowers!

May I have the CSR reviewed? https://bugs.openjdk.org/browse/JDK-8291950. The new feature deadline is approaching.  I appreciate if the CSR could be reviewed in 2 weeks.

**Comment by mcpowers** [comment]
I'll look at your CSR this morning. Make sure your SSLParameters additions look correct with JavaDoc. I always forget that step.

**Comment by XueleiFan** [comment]
> I'll look at your CSR this morning. Make sure your SSLParameters additions look correct with JavaDoc. I always forget that step.

Thank you for the suggestion.  I double checked the JavaDoc.  It looks good to me.

Please feel free to add your name as reviewer in the CSR and approve this PR when you are ready.

**Comment by mcpowers** [comment]
Unfortunately, I only have author status and can only comment.

**Comment by XueleiFan** [comment]
> Unfortunately, I only have author status and can only comment.

I think OpenJDK Author can approve as well.  I just need to get another Reviewer approval before can integration.

**Comment by seanjmullan** [comment]
> > Unfortunately, I only have author status and can only comment.
> 
> I think OpenJDK Author can approve as well. I just need to get another Reviewer approval before integration.

For a CSR, I believe that is true. But you will need a Reviewer for the PR, and they may have comments on parts that are covered by the CSR.

I would suggest moving the CSR to Proposed (which doesn't require a Reviewer AFAIK). That will move it forward a bit.

**Comment by XueleiFan** [comment]
> > > Unfortunately, I only have author status and can only comment.
> > 
> > 
> > I think OpenJDK Author can approve as well. I just need to get another Reviewer approval before integration.
> 
> For a CSR, I believe that is true. But you will need a Reviewer for the PR, and they may have comments on parts that are covered by the CSR.
> 

+1.

> I would suggest moving the CSR to Proposed (which doesn't require a Reviewer AFAIK). That will move it forward a bit.

I changed the CSR to Proposed.  Thanks for the suggestion.

**Comment by XueleiFan** [comment]
@seanjmullan Did you have cycle to review this PR and CSR?  I was wondering if it is possible to have it in JDK 20 so that more specific TLS benchmarks could be introduced sooner.  Thanks!

**Comment by seanjmullan** [comment]
> @seanjmullan Did you have cycle to review this PR and CSR? This CSR is similar to the one we did for signature algorithms. I was wondering if it is possible to have it in JDK 20 so that more specific TLS benchmarks could be introduced sooner. Thanks!

I added my name as Reviewer to the CSR, so you can finalize that, as that will give you a better chance of it being approved and getting this into JDK 20. It will take me a little while longer to review the code changes, though I don't expect it will impact the API.

File: test/jdk/javax/net/ssl/SSLParameters/NamedGroupsSpec.java
Comment (seanjmullan) [inline test/jdk/javax/net/ssl/SSLParameters/NamedGroupsSpec.java:27]: Can you write a more specific summary of what this test is testing? The @summary doesn't have to match the bug title (and often should not IMO). Since this doesn't test DTLS, that should not be included.

Same comment for the other tests.

File: test/jdk/javax/net/ssl/SSLParameters/NamedGroupsSpec.java
Comment (seanjmullan) [inline test/jdk/javax/net/ssl/SSLParameters/NamedGroupsSpec.java:34]: How about adding a test for a `null` array?

File: test/jdk/javax/net/ssl/SSLParameters/NamedGroupsSpec.java
Comment (seanjmullan) [inline test/jdk/javax/net/ssl/SSLParameters/NamedGroupsSpec.java:68]: You should also test that `getNamedGroups` returns the same elements.

File: test/jdk/javax/net/ssl/SSLParameters/NamedGroups.java
Comment (seanjmullan) [inline test/jdk/javax/net/ssl/SSLParameters/NamedGroups.java:122]: I think this case will throw IAE when `SSLParameter.setNamedGroups()` is called right? If so, I think you can delete this test (and the one one on line 123) since it is already covered in the NamedGroupsSpec test. Also, this test is coded such that the expected exception is thrown when the TLS handshake is being made in `runClientApplication()`, which is not the case here.

File: src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java
Comment (seanjmullan) [inline src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java:279]: The comment needs to be updated for named groups. It looks like it was copied from line 272.

File: src/java.base/share/classes/sun/security/ssl/NamedGroup.java
Comment (seanjmullan) [inline src/java.base/share/classes/sun/security/ssl/NamedGroup.java:454]: Add a comment describing what this method does. And the method on line 471.

Comment (XueleiFan) [reply src/java.base/share/classes/sun/security/ssl/NamedGroup.java:456]: Updated.

File: src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java
Comment (XueleiFan) [reply src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java]: Thanks for the catch.  Updated.

File: test/jdk/javax/net/ssl/SSLParameters/NamedGroups.java
Comment (XueleiFan) [reply test/jdk/javax/net/ssl/SSLParameters/NamedGroups.java:122]: It is allowed to set empty named groups with `SSLParameter.setNamedGroups()`, which means (see spec at `getNamedGroups`) that 

```
     * If the returned array is empty (zero-length), then the named group
     * negotiation mechanism is turned off for SSL/TLS/DTLS protocols, and
     * the connections may not be able to be established if the negotiation
     * mechanism is required by a certain SSL/TLS/DTLS protocol.
```

The TLS connection will fail, as expected, although for this case.

File: test/jdk/javax/net/ssl/SSLParameters/NamedGroupsSpec.java
Comment (XueleiFan) [reply test/jdk/javax/net/ssl/SSLParameters/NamedGroupsSpec.java]: It makes senses to me.  Updated accordingly.

File: test/jdk/javax/net/ssl/SSLParameters/NamedGroupsSpec.java
Comment (XueleiFan) [reply test/jdk/javax/net/ssl/SSLParameters/NamedGroupsSpec.java]: It is a good case that I missed.  Added.

File: test/jdk/javax/net/ssl/SSLParameters/NamedGroupsSpec.java
Comment (XueleiFan) [reply test/jdk/javax/net/ssl/SSLParameters/NamedGroupsSpec.java:75]: Good point.  Added.  Thank you!

**Review by seanjmullan — APPROVED** [review]
Test results are good.

**Pull request closed without merging**
